File: ios/App/Sources/Core/Domain/NavCaddy/LLM/LLMClient.swift
import Foundation

/// Client for LLM-based intent classification.
///
/// This protocol allows for different implementations (Gemini, OpenAI, mock, etc.)
/// and facilitates testing by enabling dependency injection.
///
/// Spec R2: Intent classification and entity extraction using LLM.
protocol LLMClient {
    /// Classifies user input and extracts intent with entities.
    ///
    /// - Parameters:
    ///   - input: The user's input text or transcribed speech
    ///   - context: Optional session context for conversation continuity
    /// - Returns: LLM response containing classification result
    /// - Throws: LLMError if classification fails
    func classify(input: String, context: SessionContext?) async throws -> LLMResponse
}

/// Errors that can occur during LLM operations.
enum LLMError: Error, Equatable {
    /// Network request failed
    case networkError(String)

    /// Request timed out
    case timeout

    /// Failed to parse LLM response
    case parseError(String)

    /// Invalid API key or authentication failure
    case authenticationError

    /// Rate limit exceeded
    case rateLimitExceeded

    /// Invalid input provided
    case invalidInput(String)

    /// Generic error with message
    case unknown(String)
}

extension LLMError: LocalizedError {
    var errorDescription: String? {
        switch self {
        case .networkError(let message):
            return "Network error: \(message)"
        case .timeout:
            return "Request timed out"
        case .parseError(let message):
            return "Failed to parse response: \(message)"
        case .authenticationError:
            return "Authentication failed"
        case .rateLimitExceeded:
            return "Rate limit exceeded"
        case .invalidInput(let message):
            return "Invalid input: \(message)"
        case .unknown(let message):
            return "Unknown error: \(message)"
        }
    }
}

File: android/app/src/main/java/com/example/app/domain/navcaddy/llm/LLMClient.kt
package caddypro.domain.navcaddy.llm

import caddypro.domain.navcaddy.models.SessionContext

/**
 * Interface for LLM-based intent classification.
 *
 * Abstraction layer that allows testing with mocked responses and
 * provides flexibility to switch between different LLM providers.
 *
 * Spec reference: navcaddy-engine.md R2, navcaddy-engine-plan.md Task 6
 */
interface LLMClient {
    /**
     * Classify user input into a structured intent with entities.
     *
     * @param input Raw user input (text or transcribed voice)
     * @param context Optional session context for conversational follow-ups
     * @return LLM response with classification result and metadata
     * @throws LLMException if classification fails
     */
    suspend fun classify(input: String, context: SessionContext? = null): LLMResponse
}

/**
 * Exception thrown when LLM classification fails.
 *
 * @property message Error description
 * @property cause Original throwable if available
 */
class LLMException(
    message: String,
    cause: Throwable? = null
) : Exception(message, cause)
